src/drivers/net/mcst/l_e1000.c

```diff
@@ -0,0 +1,728 @@
+	val = e1000_read_mgio_csr(ep);
+	val &= ~LSTS;
+	val |= SLSP;
+	e1000_write_mgio_csr(ep, val);
+	log_boot(" write mgio csr #0x04 (BIST_CFG2): 0x%x\n", val);
+
+	return 0;
+}
+
+static int l_e1000_open(struct net_device *dev) {
+	log_debug("mac addr: %x:%x:%x:%x:%x:%x\n", dev->dev_addr[0], dev->dev_addr[1], dev->dev_addr[2], dev->dev_addr[3], dev->dev_addr[4], dev->dev_addr[5]);
+	l_e1000_reset(dev);
+	return 0;
+}
+
+static int l_e1000_xmit(struct net_device *dev, struct sk_buff *skb) {
+	struct l_e1000_priv *ep = netdev_priv(dev);
+	uint16_t status;
+	int len;
+
+
+	if (skb->len < 0x40) {
+		len = 0x40;
+	} else {
+		len = skb->len;
+	}
+	show_packet(skb_get_data_pointner(skb->data), len, "transmit");
+
+	ep->tx_ring->base = (uint32_t)(uintptr_t)skb_get_data_pointner(skb->data);
+	ep->tx_ring->buf_length = htole16(-len);
+	ep->tx_ring->misc = 0x00000000;
+	status = TX_ST_OWN | TX_ST_ENP | TX_ST_STP;
+	ep->tx_ring->status = htole16(status);
+	wmb();
+	e1000_write_e_csr(ep, INEA|TDMD);
+	return 0;
+}
+
+static int l_e1000_set_mac_address(struct net_device *dev, const void *addr) {
+	memcpy(&dev->dev_addr[0], addr, sizeof(dev->dev_addr));
+	log_debug("mac addr: %x:%x:%x:%x:%x:%x\n", dev->dev_addr[0], dev->dev_addr[1], dev->dev_addr[2], dev->dev_addr[3], dev->dev_addr[4], dev->dev_addr[5]);
+//	l_e1000_open(dev);
+	return 0;
+}
+
+static int l_e1000_stop(struct net_device *dev) {
+	return 0;
+}
+
+static const struct net_driver l_e1000_drv_ops = {
+	.xmit = l_e1000_xmit,
+	.start = l_e1000_open,
+	.stop = l_e1000_stop,
+	.set_macaddr = l_e1000_set_mac_address
+};
+
+static int l_e1000_rx(struct l_e1000_priv *ep) {
+	uint8_t * buf = NULL;
+	int msg_len;
+	int entry;
+
+	entry  = ep->cur_rx & RX_RING_MOD_MASK;
+
+	/* If we own the next entry, it's a new packet. Send it up. */
+	while ((le16toh(ep->rx_ring[entry].status)) >= 0) {
+		struct sk_buff *skb;
+		int status;
+
+		ep->cur_rx++;
+
+		status = le16toh(ep->rx_ring[entry].status);
+		skb = ep->rx_skbs[entry];
+
+		if ((status & 0xff00) != (RD_ENP|RD_STP)) {/* There was an error. */
+			msg_len = 0;
+			if (status & RD_ENP) {
+				log_debug(" ENP");
+			} /* No detailed rx_errors counter to increment at the */
+			/* end of a packet.*/
+			if (status & RD_FRAM) {
+				log_debug(" FRAM");
+			}
+			if (status & RD_OFLO) {
+				log_debug(" OFLO ");
+			}
+	    	if (status & RD_CRC) {
+				log_debug(" CRC ");
+			}
+			if (status & RD_BUFF) {
+				log_debug(" BUFF ");
+			}
+			skb_free(skb);
+		} else {
+			buf = (void *)(uintptr_t)le32toh(ep->rx_ring[entry].base);
+			msg_len =  (le16toh(ep->rx_ring[entry].msg_length) & 0xfff) - ETH_FCS_LEN;
+			show_packet(buf, msg_len, " --- BUF received packet print:");
+			skb->dev = ep->netdev;
+			skb->len = msg_len;
+
+			netif_rx(skb);
+		}
+
+		ep->rx_skbs[entry] = l_e1000_new_rx_buff(ep, entry);
+
+		entry  = ep->cur_rx & RX_RING_MOD_MASK;
+	}
+	return 0;
+}
+
+static irq_return_t l_e1000_interrupt(unsigned int irq_num, void *dev_id) {
```

Comment:
I do not understand where we clear (and if it is required or not) interrupts in this function. On one hand, I see `csr0 &= ~BABL;` but we do not use csr0 value to write back to the register back. On the other hand, I do not see `csr0 &= ~TINT;` and `csr0 &= ~RINT;`. Can you please clarify these moments?

Reply:
Ah, sorry I see `e1000_write_e_csr(ep, csr0|IDON);`

Reply:
BTW, then you can safely remove `csr0 &= ~BABL;` and others same things, to make this place more clear.

Reply:
yep, I agree. Done.